AppDynamics plugin (#674)
Kong/developer.konghq.com: app/_kong_plugins/app-dynamics/index.md

icon: app-dynamics.png

categories:
  - analytics

>   - monitoring

Reviewer: ```suggestion
  - analytics-monitoring
  
tags:
  - analytics
  - monitoring
```
The category expects `analytics-monitoring` to sort the plugins in the right place in the plugin hub: https://deploy-preview-674--kongdeveloper.netlify.app/plugins/?category=analytics-monitoring

You can add these add individual tags though.

>   - analytics
>   - monitoring

Author: ```suggestion
  - analytics-monitoring
```

- If {{site.base_gateway}} is deployed in RHEL or CentOS, the `libappdynamics.so` file can be in the `/usr/lib64` directory, which is included in the default search path for shared libraries.
- If {{site.base_gateway}} is deployed in Debian or Ubuntu, the `libappdynamics.so` file can be in the `/usr/lib` directory, which is included in the default search path for shared libraries.
- If above options are not available, the `libappdynamics.so` file can be in one of the locations configured by the [system's shared library loader](https://tldp.org/HOWTO/Program-Library-HOWTO/shared-libraries.html).
- Alternatively, the `LD_LIBRARY_PATH` environment variable can be set to the directory containing the `libappdynamics.so` file when starting {{site.base_gateway}}.

## Enable the AppDynamics plugin

Before you configure the plugin, you must enable it:

* **Package install:** Set `plugins=bundled,app-dynamics` in [`kong.conf`](/gateway/configuration/) before starting {{site.base_gateway}}
* **Docker:** Set `export KONG_PLUGINS=bundled,app-dynamics` in the environment

> * **Kubernetes:** Set `KONG_PLUGINS=bundled,app-dynamics` using these instructions

Reviewer: What are "these instructions"?

## Plugin configuration

The AppDynamics plugin is configured through environment variables
that must be set when {{site.base_gateway}} is started. The AppDynamics plugin makes use of the AppDynamics C/C++ SDK to send
information to the AppDynamics controller. See the
[AppDynamics C/C++ SDK documentation](https://docs.appdynamics.com/appd/23.x/latest/en/application-monitoring/install-app-server-agents/c-c++-sdk/use-the-c-c++-sdk)
for more information about the configuration parameters.

{:.info}

> > All non-default environment variables in the table **must** be set.

Reviewer: Can we make them `required` instead, and label them that way?

| `KONG_APPD_APP_NAME` | Application name to report to AppDynamics. | String | `Kong` |
| `KONG_APPD_NODE_NAME` | Node name to report to AppDynamics. This value defaults to the system's hostname.| String | `hostname` |
| `KONG_APPD_INIT_TIMEOUT_MS` | Maximum time to wait for a controller connection when starting, in milliseconds. | Integer | `5000` |
| `KONG_APPD_CONTROLLER_USE_SSL` | Use SSL encryption in controller communication. `true`, `on`, or `1` are all interpreted as `True`, any other value is considered `false`.| Boolean | `on` |
| `KONG_APPD_CONTROLLER_HTTP_PROXY_HOST` | Hostname of proxy to use to communicate with controller. | String |  |
| `KONG_APPD_CONTROLLER_HTTP_PROXY_PORT` | Port number of controller proxy. | Integer |  |
| `KONG_APPD_CONTROLLER_HTTP_PROXY_USERNAME` | Username to use to identify to proxy. This value is a string that is never shown in logs. This value can be specified as a vault reference.| String |  |
| `KONG_APPD_CONTROLLER_HTTP_PROXY_PASSWORD` | Password to use to identify to proxy. This value is a string that is never shown in logs. This value can be specified as a vault reference.| String |  |

> | `KONG_APPD_CONTROLLER_CERTIFICATE_FILE` | Path to a self-signed certificate file. For example, `/etc/kong/certs/ca-certs.pem`. <br><br>_Available starting in {{site.base_gateway}} 3.4.3.3_ | String | | 
> | `KONG_APPD_CONTROLLER_CERTIFICATE_DIR` | Path to a certificate directory. For example, `/etc/kong/certs/`. <br><br> _Available starting in {{site.base_gateway}} 3.4.3.3_ | String | | 
> | `KONG_APPD_CONTROLLER_CERTIFICATE_FILE` | Path to a self-signed certificate file. For example, `/etc/kong/certs/ca-certs.pem`. <br><br>_Available starting in {{site.base_gateway}} 3.6.x_ | String | | 
> | `KONG_APPD_CONTROLLER_CERTIFICATE_DIR` | Path to a certificate directory. For example, `/etc/kong/certs/`. <br><br>_Available starting in {{site.base_gateway}} 3.6.x_ | String | | 

Reviewer: ```suggestion
| `KONG_APPD_CONTROLLER_CERTIFICATE_FILE` | Path to a self-signed certificate file. For example, `/etc/kong/certs/ca-certs.pem`. <br><br>_Available starting in {{site.base_gateway}} 3.4.3.3_ | String | | 
| `KONG_APPD_CONTROLLER_CERTIFICATE_DIR` | Path to a certificate directory. For example, `/etc/kong/certs/`. <br><br> _Available starting in {{site.base_gateway}} 3.4.3.3_ | String | | 
```
I can't see any difference between these. I'm guessing it was probably a backport that created some dupe table entries/version tags.

| `KONG_APPD_ANALYTICS_ENABLE` | Enable or disable Analytics Agent reporting. When disabled (default), Analytics-related logging messages are suppressed. <br><br>_Available starting in {{site.base_gateway}} 3.8.x_ | Boolean | `false` | 

### Possible values for the `KONG_APPD_LOGGING_LEVEL` parameter

The `KONG_APPD_LOGGING_LEVEL` environment variable is a numeric value that controls the desired logging level.
Each value corresponds to a specific level:

| Value | Name | Description |
|--|--|--|
| 0 | `TRACE` | Reports finer-grained informational events than the debug level that may be useful to debug an application. |
| 1 | `DEBUG` | Reports fine-grained informational events that may be useful to debug an application. |
| 2 | `INFO` | Default log level. Reports informational messages that highlight the progress of the application at coarse-grained level.|
| 3 | `WARN` | Reports on potentially harmful situations. |
| 4 | `ERROR` | Reports on error events that may allow the application to continue running.|
| 5 | `FATAL` | Fatal errors that prevent the agent from operating. |

## Agent logging

> The AppDynamics agent sorts log information into separate log files, independent of {{site.base_gateway}}.

Reviewer: ```suggestion
The AppDynamics agent sorts log information into separate log files, independent of {{site.base_gateway}} logs.
```

By default, log files are written to the `/tmp/appd` directory.
This location can be changed by setting the `KONG_APPD_LOGGING_LOG_DIR` environment variable.

If problems occur with the AppDynamics integration, inspect the AppDynamics agent's log files in addition to the {{site.base_gateway}} logs.

## AppDynamics node name considerations

The AppDynamics plugin sets the `KONG_APPD_NODE_NAME` to the local

> host name by default, which typically reflects the container ID of the containerized

Reviewer: ```suggestion
hostname by default, which typically reflects the container ID of the containerized
```

application. Multiple instances of the AppDynamics agent must use

> different node names, and one agent must exists for each of {{site.base_gateway}}'s

Reviewer: ```suggestion
different node names, and one agent must exist for each of {{site.base_gateway}}'s
```

  - app-dynamics
---

This plugin integrates {{site.base_gateway}} with the [AppDynamics APM platform](https://www.splunk.com/en_us/products/splunk-appdynamics.html) so that
proxy requests handled by {{site.base_gateway}} can be identified and analyzed in
AppDynamics. 

The plugin reports request and response timestamps and error information to 
the AppDynamics platform to be analyzed in the AppDynamics flow map and correlated 
with other systems participating in handling application API requests.

{:.warning}

> > **Important:** Unlike other {{site.base_gateway}} plugins, you must configure the AppDynamics plugin via environment variables. You must also install AppDynamics before using the plugin.

Reviewer: ```suggestion
> **Important:** Unlike other {{site.base_gateway}} plugins, you must configure the AppDynamics plugin via environment variables. You must also install AppDynamics before using the plugin, and you must enable the plugin in your environment.
```

This directory is included in the {{site.base_gateway}} search path for shared libraries, so the `libappdynamics.so` file will be found automatically.

### Alternative installation

If you are using an older version of {{site.base_gateway}}, or if you prefer to install the `libappdynamics.so` file in a different location, you can do so.

- If {{site.base_gateway}} is deployed in RHEL or CentOS, the `libappdynamics.so` file can be in the `/usr/lib64` directory, which is included in the default search path for shared libraries.
- If {{site.base_gateway}} is deployed in Debian or Ubuntu, the `libappdynamics.so` file can be in the `/usr/lib` directory, which is included in the default search path for shared libraries.
- If above options are not available, the `libappdynamics.so` file can be in one of the locations configured by the [system's shared library loader](https://tldp.org/HOWTO/Program-Library-HOWTO/shared-libraries.html).
- Alternatively, the `LD_LIBRARY_PATH` environment variable can be set to the directory containing the `libappdynamics.so` file when starting {{site.base_gateway}}.

> ## Enable the AppDynamics plugin
> 
> Before you configure the plugin, you must enable it:

Reviewer: ```suggestion
## Enable the AppDynamics plugin

The AppDynamics plugin is not bundled in {{site.base_gateway}} packages by default. 
Before you configure the plugin, you must enable it:
```
